src/v/kafka/server/handlers/handler_probe.cc

```diff
@@ -0,0 +1,123 @@
+ */
+
+#include "kafka/server/handlers/handler_probe.h"
+
+#include "config/configuration.h"
+#include "kafka/server/handlers/handler_interface.h"
+#include "prometheus/prometheus_sanitize.h"
+
+#include <seastar/core/lowres_clock.hh>
+#include <seastar/core/metrics.hh>
+
+#include <chrono>
+#include <memory>
+
+namespace kafka {
+
+handler_probe_manager::handler_probe_manager()
+  : _probes(max_api_key() + 2)
+  , _unknown_handler_index(max_api_key() + 1) {
+    for (size_t i = 0; i < _probes.size(); i++) {
+        auto key = api_key{i};
+
+        if (handler_for_key(key) || i == _unknown_handler_index) {
+            _probes[i] = std::make_unique<handler_probe>(key);
+            _probes.at(i)->setup_metrics();
+        }
+    }
+}
+
+handler_probe& handler_probe_manager::get_probe(api_key key) {
+    if (!handler_for_key(key)) {
+        key = api_key{_unknown_handler_index};
+    }
+
+    return *_probes.at(key);
+}
+
+handler_probe::handler_probe(api_key key)
+  : _key(key) {}
+
+void handler_probe::setup_metrics() {
+    namespace sm = ss::metrics;
+
+    if (config::shard_local_cfg().disable_metrics()) {
+        return;
+    }
+
+    const char* handler_name;
+    if (auto handler = handler_for_key(_key)) {
+        handler_name = handler.value()->name();
+    } else {
+        handler_name = "unknown_handler";
+    }
+
+    std::vector<sm::label_instance> labels{sm::label("handler")(handler_name)};
+    auto aggregate_labels = config::shard_local_cfg().aggregate_metrics()
+                              ? std::vector<sm::label>{sm::shard_label}
+                              : std::vector<sm::label>{};
+
+    _metrics.add_group(
+      prometheus_sanitize::metrics_name("kafka:handler"),
+      {
+        sm::make_counter(
+          "requests_completed",
+          [this] { return _requests_completed; },
+          sm::description("Number of requests completed"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_errored",
+          [this] { return _requests_errored; },
+          sm::description("Number of requests errored"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "requests_in_progress",
+          [this] { return _requests_in_progress_every_ns / 1'000'000'000; },
+          sm::description("Number of requests in progress"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_received",
+          [this] { return _bytes_received; },
+          sm::description("Number of bytes received from requests"),
+          labels)
+          .aggregate(aggregate_labels),
+        sm::make_counter(
+          "bytes_sent",
+          [this] { return _bytes_sent; },
+          sm::description("Number of bytes sent in replies"),
+          labels)
+          .aggregate(aggregate_labels),
+      });
+}
+
+/*
+ * This roughly approximates an integral of `_requests_in_progress`.
```

Comment:
Why is it rough? It should be exact, no?

Reply:
Oh is it because of the coarse granularity of lowres_clock?

Reply:
Yeah, a bit because of the coarse granularity of the lowres_clock. A bit because you really can't have a true integral over a discrete function. And a bit because I wrote this back when I was sampling at 1ms intervals lol.

Reply:
There's no problem in taking the integral (or at least the value we care about) over this function even though it is discontinuous. I guess in math terms you'd say its points of discontinuity have "measure zero" so it is [Riemann-integrable](https://en.wikipedia.org/wiki/Riemann_integral).

So I see it as exact modulo any inaccuracy in the clock (but that applies to all of our other timings here too).